.equals() and .ionEquals() for dom.Value equivalence

desaikd opened the pull request:
*Issue #681*

*Description of changes:*
This PR works on creating two methods for `dom.Value` equivalence:

* `ionEquals`: For checking equivalence of two `dom.Value` instances
* `equals`: For checking non-strict equivalence of a `dom.Value` and another value. The other value may be a `dom.Value` or a native JS type.

*Completed Tasks:*
- Created renamed equals method as `ionEquals` to compare with `dom.Value` only. It contains options for strict/non-strict comparisons. 
- Remove `onlyCompareIon` option entirely from `ionEquals`.
- Created `equals` for JS Object equivalence with Ion Value. Allows only non-strict comparison. 
- Renamed `_ionEquals` helper method as `_valueEquals`
- Tests suite changes for using `equals` / `ionEquals` accordingly for tests.
- Removed `onlyCompareIon` option from namespace `Value.Equality`.
- Both `Sequence` and `Lob` equivalence changed to use `ionEquals` or `equals` based on `onlyCompareIon` for child value comparisons.

*Questions:*
1.  Any more tests that should be added here? 

*Equivalence reference:*

**Strict** equivalence refers to Ion data model equivalence as defined in Ion Equivalence[1] and by Ion Specification[2]

**Structural** or **non-strict** equivalence follows the same rules as strict equivalence, except that
1. Annotations are not considered, and
2. Timestamps that represent the same instant in time are always
     considered equivalent.
     
[1] https://www.javadoc.io/doc/com.amazon.ion/ion-java/latest/com/amazon/ion/util/Equivalence.html
[2]  http://amzn.github.io/ion-docs/docs/spec.html

zslayton commented on src/dom/Value.ts:
```diff
@@ -206,18 +206,21 @@ export interface Value {
   ): boolean;
+
+  /**
+   * For checking non-strict equivalence of an Ion Values represented by dom.Value with the other value represented
```
```suggestion
   * Compares this instance of dom.Value to the provided value and returns true
   * if they are equal. If `other` is a dom.Value, this comparison checks for structural
   * (or "non-strict") equivalence. If `other` is a native JS value, this comparison will
   * convert it to an Ion value and then check for structural equivalence.
```

zslayton replied:
Sorry, I just re-read the implementation for `_valueEquals` and remembered it doesn't convert JS types to Ion types during the comparison. That part of my suggested edit wouldn't be accurate.

zslayton commented on test/dom/equivalence.ts:
```diff
@@ -74,41 +74,41 @@ describe("Equivalence", () => {
+    assert.isTrue(float1.equals(1.5));
+    assert.isFalse(float1.ionEquals(1.5));
+    assert.isFalse(float1.equals(1.2));
+    // assert.isTrue(float1.equals(1.2, {epsilon: 0.5}));
+    assert.isTrue(float1.equals(new Number(1.5)));
+    assert.isFalse(float1.equals(new Number(1.2)));
+    // assert.isTrue(
+    //     float1.equals(new Number(1.2), {epsilon: 0.5})
+    // );
```
Some leftover comments here.

desaikd commented:
*Changes:*

1. Made `ionEquals` accept only `dom.Value` type for comparison where comparison type can be `non strict` or `strict`.
2. changed `equals` to accept either `dom.Value` or JS object for comparison. comparison type can be only `non strict`.
3. added epsilon option into `equals`. 
4. Changed `Decimal` equivalence method `_valueEquals` to accept `Float` and compare by converting it to `Decimal`.
5. Changes in test suite for `epsilon` option demonstration in `equals`.

zslayton commented on src/dom/Value.ts:
```diff
@@ -217,10 +217,21 @@ export interface Value {
+   *
+   * @param other                       other Ion Value or Js Object to be compared with this Ion Value.
+   * @param options                     options provided for equivalence as below
+   *        epsilon                     used by Float for an equality with given epsilon precision. (Default: null)
```
Now that `equals` allows for comparisons across numeric types (e.g. `someFloat.equals(someDecimal)`), should `epsilon` has to be considered across types too?

I'm inclined to say yes for consistency, but `ion.Decimal` doesn't currently make it very easy to do arithmetic so it might take non-trivial effort to implement.

zslayton commented on src/dom/Value.ts:
```diff
@@ -217,10 +217,21 @@ export interface Value {
+   * Compares this instance of dom.Value to the provided value and returns true
+   * if they are equal. If `other` is a dom.Value, this comparison checks for structural
+   * (or "non-strict") equivalence. If `other` is a native JS value, this comparison will
+   * convert it to an Ion value and then check for structural equivalence.
```
> If `other` is a native JS value, this comparison will convert it to an Ion value and then check for structural equivalence.

I know this was originally my phrasing suggestion, but I think we determined that it's not accurate. We do the comparisons without converting the JS value to Ion first.

desaikd commented:
*Changes:*
1. Added support `Float` and `Decimal` comparisons when using `equals`.
2. Remove `Value.Equality` namespace.
3. Tests suite changes for removing `Value.Equality`.
4. added `equals` option to `_valueEquals` private helper method.

zslayton commented on src/dom/Boolean.ts:
```diff
@@ -75,13 +75,12 @@ export class Boolean extends Value(
-        valueToCompare = other.booleanValue();
-      }
-    } else {
+    // `compareOnlyIon` requires that the provided value be an ion.dom.Boolean instance.
```
Now that we're not checking `if (options.onlyCompareIon)` here, I think this comment may be confusing. (Same thought in the other similarly refactored areas.)

zslayton commented on src/dom/Boolean.ts:
```diff
@@ -75,13 +75,12 @@ export class Boolean extends Value(
+      isSupportedType = true;
+      valueToCompare = other.booleanValue();
+    }
+    if (!options.onlyCompareIon) {
```
This could be an `else if`, I think?

zslayton commented on src/dom/Decimal.ts:
```diff
@@ -66,26 +65,27 @@ export class Decimal extends Value(
+      isSupportedType = true;
+      valueToCompare = new ion.Decimal(other.toString());
+    }
+    if (!options.onlyCompareIon) {
```
I think these can be `else if`s too. (I'll stop adding this comment from here on.)

zslayton reviewed (approved):
Looks good! Some minor cleanup suggestions below.
